### PR TITLE
[Clang] allow restrict qualifier for array types with pointer types as element types

### LINKED ISSUE #92847: C23 restrict-on-array rejected
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf page 120 (6.7.4 Type qualifiers, Constraints):
> Types other than pointer types whose referenced type is an object  type and (possibly multi-dimensional) array types with such pointer types as element type shall not be restrict-qualified.

Thus: `int *` can be restricted, as can `(int *)[10]` and `(int *)[10][10]`.

Clang agrees with the former, and
```c
int *restrict x[1];
```
works, but
```c
typedef int *t[1];
restrict t x;
```
is rejected as
```
$ cc -std=c23 a.c -c
a.c:2:1: error: restrict requires a pointer or reference ('t' (aka 'int *[1]') is invalid)
    2 | restrict t x;
      | ^
```

Debian clang version 19.0.0 (++20240422112300+a2692ac23f14-1\~exp1\~20240422112407.665).

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #92847

---

> Types other than pointer types whose referenced type is an object type and (possibly multi-dimensional) array types with such pointer types as element type shall not be restrict-qualified.

### CODE REVIEW

**Comment by cor3ntin** [comment]
@AaronBallman ping

**Comment by a-tarasyuk** [comment]
@AaronBallman could you review these changes? thanks

**Comment by a-tarasyuk** [comment]
@AaronBallman could you review these changes? thanks

File: clang/test/Sema/types.c [hunk 1]
```diff
@@ -9,20 +9,20 @@ typedef int (*T)[2];
 restrict T x;
 
 typedef int *S[2];
-restrict S y; // expected-error {{restrict requires a pointer or reference ('S' (aka 'int *[2]') is invalid)}}
-
-
+restrict S y;
```
Comment (AaronBallman) [inline clang/test/Sema/types.c:12]: Can you also add a test for multidimensional arrays, like:
```
typedef int *S[2][2];
restrict S y;
```
and (this should already work because it's adding the restrict to a pointer to an array):
```
typedef int (*S)[2];
restrict S y;
```
and finally, can you add some type verification tests like (for each of the test cases):
```
static_assert(_Generic(typeof(y), int * restrict[2]: 1, default : 0));
```

**Review by AaronBallman — COMMENTED** [review]
I believe these changes are correct, but adding a few more folks interested in C to verify because existing compilers definitely disagree: https://godbolt.org/z/1Yb44WP9o

File: clang/lib/Sema/SemaType.cpp [hunk 2]
```diff
@@ -1596,12 +1596,14 @@ QualType Sema::BuildQualifiedType(QualType T, SourceLocation Loc,
     QualType ProblemTy;
 
     if (T->isAnyPointerType() || T->isReferenceType() ||
-        T->isMemberPointerType()) {
+        T->isMemberPointerType() || T->isArrayType()) {
```
Comment (efriedma-quic) [inline clang/lib/Sema/SemaType.cpp:1599]: Don't you need some kind of recursion to check that the base element type ?  We don't want to allow `typedef int x[2]; typedef x restrict x2;`.  Or `typedef int (*x)(); typedef x x2[2]; typedef x2 restrict x3;`

File: clang/lib/Sema/SemaType.cpp [hunk 3]
```diff
@@ -1596,12 +1596,14 @@ QualType Sema::BuildQualifiedType(QualType T, SourceLocation Loc,
...
       QualType EltTy;
       if (T->isObjCObjectPointerType())
         EltTy = T;
       else if (const MemberPointerType *PTy = T->getAs<MemberPointerType>())
         EltTy = PTy->getPointeeType();
+      else if (T->isArrayType())
+        EltTy = BuildQualifiedType(Context.getBaseElementType(T), Loc, Qs);
```
Comment (efriedma-quic) [inline clang/lib/Sema/SemaType.cpp:1606]: I'm not sure this is the best way to write this... I mean, I guess it works, but it's hard to follow.  Can you separate out the array handling a bit more?  Maybe just put `QualType ProblemTy = Context.getBaseElementType(T)` before you start checking the inner type.

File: clang/test/Sema/types.c (hunk 1 above)
Comment (a-tarasyuk) [reply clang/test/Sema/types.c:12]: @AaronBallman thanks for the feedback. I've added additional tests

File: clang/lib/Sema/SemaType.cpp (hunk 3 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp]: @efriedma-quic I've moved `getBaseElementType`

Comment (efriedma-quic) [reply clang/lib/Sema/SemaType.cpp]: I guess I wasn't clear enough.

My suggestion was to call `Context.getBaseElementType(T)`, and then perform all the pointer type/object type/etc. on the result of that, instead of recursively calling BuildQualifiedType.

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp]: @efriedma-quic Whoops :). I had considered removing recursion but was not certain that was the main point. I've removed recursion. Could you check the latest changes? thanks

File: clang/lib/Sema/SemaType.cpp [hunk 4]
```diff
@@ -1593,34 +1593,31 @@ QualType Sema::BuildQualifiedType(QualType T, SourceLocation Loc,
...
-    if (T->isAnyPointerType() || T->isReferenceType() ||
-        T->isMemberPointerType()) {
-      QualType EltTy;
-      if (T->isObjCObjectPointerType())
-        EltTy = T;
-      else if (const MemberPointerType *PTy = T->getAs<MemberPointerType>())
+    QualType EltTy = Context.getBaseElementType(T);
+
+    if (EltTy->isAnyPointerType() || EltTy->isReferenceType() ||
+        EltTy->isMemberPointerType() || EltTy->isArrayType()) {
```
Comment (efriedma-quic) [inline clang/lib/Sema/SemaType.cpp:1599]: Drop the isArrayType() here; should be impossible for EltTy to be an array here.

File: clang/test/Sema/restrict-qualifier.c [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+restrict T1 t1;
+static_assert(_Generic(typeof (t1), int (*restrict)[2] : 1, default : 0));
+
+typedef int *T2[2];
+restrict T2 t2;
+static_assert(_Generic(typeof (t2), int *restrict[2] : 1, default : 0));
+
+typedef int *T3[2][2];
+restrict T3 t3;
+static_assert(_Generic(typeof (t3), int *restrict[2][2] : 1, default : 0));
```
Comment (efriedma-quic) [inline clang/test/Sema/restrict-qualifier.c:13]: Maybe also add a check that we handle this correctly in function types:

```
static_assert(_Generic(void(T3 restrict), void(int *restrict(*)[2]):1, default:0));
```

Comment (a-tarasyuk) [reply clang/test/Sema/restrict-qualifier.c:13]: @AaronBallman @efriedma-quic could you review the latest changes? thanks

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Review by efriedma-quic — CHANGES_REQUESTED** [review]
Oh, just thought of one more thing: we probably want a c23_compat warning for this (assuming the language is new in c23).

**Comment by AaronBallman** [comment]
> Oh, just thought of one more thing: we probably want a c23_compat warning for this (assuming the language is new in c23).

Good catch -- this is new in C23. C17 said: "Types other than pointer types whose referenced type is an object type shall not be restrict-qualified."

So there's two diagnostics needed: one is an `Extension` warning (so it fires in `-pedantic`) for use of an extension pre-C23, and the other is `Warning`, `DefaultIgnore` for code which is incompatible with standards before C23. (And test coverage for both).

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 6]
```diff
@@ -7409,6 +7409,11 @@ def warn_c23_compat_utf8_string : Warning<
 def note_cxx20_c23_compat_utf8_string_remove_u8 : Note<
   "remove 'u8' prefix to avoid a change of behavior; "
   "Clang encodes unprefixed narrow string literals as UTF-8">;
+def warn_c23_compat_restrict_pointers_to_array : Warning<
+  "'restrict' qualifier on pointers to arrays is incompatible with C17 and earlier">,
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: @AaronBallman Thanks for the feedback. I've updated diagnostics

File: clang/test/Sema/pre-c2x-restrict-qualifier.c [hunk 7]
```diff
@@ -0,0 +1,26 @@
+// RUN: %clang_cc1 -std=c23 -fsyntax-only -verify %s
+// RUN: %clang_cc1 -std=c17 -fsyntax-only -pedantic -verify=pedantic %s
+// RUN: %clang_cc1 -std=c17 -fsyntax-only -Wpre-c2x-compat -verify=pre-c2x-compat %s
+
+typedef int (*T1)[2];
+restrict T1 t1;         // pedantic-warning {{'restrict' qualifier on pointers to arrays is a C23 extension}} \
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/test/Sema/pre-c2x-restrict-qualifier.c]: @AaronBallman I've updated test/handling so it no longer emits a warning for this case.

File: clang/test/Sema/pre-c2x-restrict-qualifier.c [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+                        // pre-c2x-compat-warning {{'restrict' qualifier on pointers to arrays is incompatible with C17 and earlier}}
+
+typedef int *T2[2];
+restrict T2 t2;         // pedantic-warning {{'restrict' qualifier on pointers to arrays is a C23 extension}} \
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/test/Sema/pre-c2x-restrict-qualifier.c]: @AaronBallman definitely, I've updated the diagnostic.

File: clang/lib/Sema/SemaType.cpp [hunk 9]
```diff
@@ -1593,35 +1593,38 @@ QualType Sema::BuildQualifiedType(QualType T, SourceLocation Loc,
...
+    Loc = DS ? DS->getRestrictSpecLoc() : Loc;
     if (DiagID) {
-      Diag(DS ? DS->getRestrictSpecLoc() : Loc, DiagID) << ProblemTy;
+      Diag(Loc, DiagID) << EltTy;
       Qs.removeRestrict();
+    } else {
+      if (!getLangOpts().C23 && T->isArrayType()) {
+        Diag(Loc, diag::ext_restrict_on_array_of_pointers_c23);
+        Diag(Loc, diag::warn_c23_compat_restrict_on_array_of_pointers);
+      }
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp]: @efriedma-quic thanks. I've added chaneges.

File: clang/test/Sema/types.c [hunk 10]
```diff
@@ -9,20 +9,20 @@ typedef int (*T)[2];
...
-}
+} // expected-warning {{non-void function does not return a value}}
+
 // but not a keyword
 int b(void) {
   int __int128_t;
   int __uint128_t;
-}
+} // expected-warning {{non-void function does not return a value}}
+
```
Comment (AaronBallman) [inline clang/test/Sema/types.c:25]: For these tests, I'd say it's better to change the `int` return type to `void` to silence the warning; neither of the tests are testing the behavior of the return anyway.

File: clang/lib/Sema/SemaType.cpp [hunk 11]
```diff
@@ -1593,35 +1593,40 @@ QualType Sema::BuildQualifiedType(QualType T, SourceLocation Loc,
...
     if (DiagID) {
-      Diag(DS ? DS->getRestrictSpecLoc() : Loc, DiagID) << ProblemTy;
+      Diag(Loc, DiagID) << EltTy;
       Qs.removeRestrict();
+    } else {
+      if (T->isArrayType()) {
+        if (getLangOpts().C23)
+          Diag(Loc, diag::warn_c23_compat_restrict_on_array_of_pointers);
+        else
+          Diag(Loc, diag::ext_restrict_on_array_of_pointers_c23);
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaType.cpp:1628]:
```suggestion
        Diag(Loc, getLangOpts().C23 ? diag::warn_c23_compat_restrict_on_array_of_pointers : diag::ext_restrict_on_array_of_pointers_c23);
```
And wrapped properly for clang-format.

File: clang/lib/Sema/SemaType.cpp [hunk 12]
```diff
@@ -1593,35 +1593,40 @@ QualType Sema::BuildQualifiedType(QualType T, SourceLocation Loc,
...
-      QualType EltTy;
-      if (T->isObjCObjectPointerType())
-        EltTy = T;
-      else if (const MemberPointerType *PTy = T->getAs<MemberPointerType>())
+    QualType EltTy = Context.getBaseElementType(T);
+
+    if (EltTy->isAnyPointerType() || EltTy->isReferenceType() ||
+        EltTy->isMemberPointerType()) {
+
+      if (const MemberPointerType *PTy = EltTy->getAs<MemberPointerType>())
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaType.cpp:1601]:
```suggestion
      if (const auto *PTy = EltTy->getAs<MemberPointerType>())
```

File: clang/test/Sema/pre-c2x-restrict-qualifier.c [hunk 13]
```diff
@@ -0,0 +1,25 @@
+// RUN: %clang_cc1 -std=c23 -fsyntax-only -verify %s
+// RUN: %clang_cc1 -std=c17 -fsyntax-only -pedantic -verify=pedantic %s
+// RUN: %clang_cc1 -std=c2x -fsyntax-only -Wpre-c2x-compat -verify=c2x-compat %s
```
Comment (AaronBallman) [inline clang/test/Sema/pre-c2x-restrict-qualifier.c:3]:
```suggestion
// RUN: %clang_cc1 -std=c23 -fsyntax-only -verify %s
// RUN: %clang_cc1 -std=c17 -fsyntax-only -pedantic -verify=pedantic,expected %s
// RUN: %clang_cc1 -std=c2x -fsyntax-only -Wpre-c2x-compat -verify=c2x-compat,expected %s
```

File: clang/test/Sema/pre-c2x-restrict-qualifier.c [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+typedef int *T3[2][2];
+restrict T3 t3;         // pedantic-warning {{'restrict' qualifier on an array of pointers is a C23 extension}} \
+                        // c2x-compat-warning {{'restrict' qualifier on an array of pointers is incompatible with C standards before C23}}
+
+typedef int (*t4)();    // pedantic-warning {{a function declaration without a prototype is deprecated in all versions of C}}
+typedef t4 t5[2];
+typedef t5 restrict t6; // expected-error {{pointer to function type 'int (void)' may not be 'restrict' qualified}} \
+                        // pedantic-error {{pointer to function type 'int ()' may not be 'restrict' qualified}} \
+                        // c2x-compat-error {{pointer to function type 'int (void)' may not be 'restrict' qualified}}
```
Comment (AaronBallman) [inline clang/test/Sema/pre-c2x-restrict-qualifier.c:20]:
```suggestion
typedef t5 restrict t6; // expected-error {{pointer to function type}}
```
(Could use expected-error-re and use a regular expression to handle the `()` vs `(void)` differences, too, but that makes the test slower to run and maybe this is sufficiently clear?)

File: clang/test/Sema/pre-c2x-restrict-qualifier.c [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+typedef int t7[2];
+typedef t7 restrict t8; // expected-error {{restrict requires a pointer or reference ('t7' (aka 'int[2]') is invalid)}} \
+                        // pedantic-error {{restrict requires a pointer or reference ('t7' (aka 'int[2]') is invalid)}} \
+                        // c2x-compat-error {{restrict requires a pointer or reference ('t7' (aka 'int[2]') is invalid}}
```
Comment (AaronBallman) [inline clang/test/Sema/pre-c2x-restrict-qualifier.c:25]:
```suggestion
typedef t7 restrict t8; // expected-error {{restrict requires a pointer or reference ('t7' (aka 'int[2]') is invalid)}}
```

**Review by AaronBallman — COMMENTED** [review]
I think this is pretty close, mostly just nits from me.

File: clang/test/Sema/types.c (hunk 10 above)
Comment (a-tarasyuk) [reply clang/test/Sema/types.c]: @AaronBallman thanks, I've made changes to all the requested cases.

**Review by AaronBallman — APPROVED** [review]
LGTM! Please wait a bit before landing in case @efriedma-quic has any additional comments.

**Review by efriedma-quic — APPROVED** [review]
